python/ray/serve/tests/unit/test_pow_2_replica_scheduler.py

```diff
@@ -1627,5 +1628,67 @@ async def test_backoff_index_handling(pow_2_scheduler, backoff_index: int):
     assert r in [r1, r2]
 
 
+@pytest.mark.asyncio
+@pytest.mark.parametrize("pow_2_scheduler", [{}], indirect=True)
+async def test_replicas_actor_died_error(
+    pow_2_scheduler: PowerOfTwoChoicesReplicaScheduler,
+):
+    """
+    If replicas return an ActorDiedError, they should be removed from the
+    local list. If they an ActorUnavailableError, they should remain in the
+    local list.
+    """
+    s = pow_2_scheduler
+
+    r1 = FakeReplicaWrapper("r1")
+    r1.set_queue_len_response(
+        queue_len=0,
+        exception=ActorDiedError(),
+    )
+
+    r2 = FakeReplicaWrapper("r2")
+    r2.set_queue_len_response(0)
+
+    s.update_replicas([r1, r2])
+
+    for _ in range(10):
+        assert (await s.choose_replica_for_request(fake_pending_request())) == r2
+
+    # The scheduler should remove r1 since it died.
+    assert set(pow_2_scheduler.curr_replicas.values()) == {r2}
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("pow_2_scheduler", [{}], indirect=True)
+async def test_replicas_actor_unavailable_error(
+    pow_2_scheduler: PowerOfTwoChoicesReplicaScheduler,
+):
+    """
+    If they an ActorUnavailableError, they should remain in the local list.
+    """
+    s = pow_2_scheduler
+
+    r1 = FakeReplicaWrapper("r1")
+    r1.set_queue_len_response(1)
+    r1.set_queue_len_response(
+        queue_len=0,
+        exception=ActorUnavailableError(
+            error_message="Actor is temporarily unavailable",
+            actor_id=b"a" * 16,
+        ),
+    )
+
+    r2 = FakeReplicaWrapper("r2")
+    r2.set_queue_len_response(0)
+
+    s.update_replicas([r1, r2])
+
+    for _ in range(10):
+        assert (await s.choose_replica_for_request(fake_pending_request())) == r2
+
+    # The scheduler should keep r1 since it may recover.
+    assert set(pow_2_scheduler.curr_replicas.values()) == {r1, r2}
```

Comment:
let's test the e2e behavior instead of verifying what are basically implementation details -- swap `r1` to stop raising the exception and check it can be scheduled to

Reply:
Sounds good, I checked that `r1` gets scheduling requests after it recovers.